### ISSUE TITLE
backward compatibility broken on vcvars command

### ISSUE THREAD

**Issue by claasd** [conan-io/conan#3917]
Since implementing #3157 in conan 1.7, using vcvars with Visual Studio 2015 and architecture 86 uses amd64_x86 as parameter for vcvars, even if arch_build is not specified and not part of the recipes settings. This breaks for example builds where Silverlight is used, as Silverlight cannot be used with 64 bit msbuild commands.

Expected behavior: if arch is set to 86 and arch_build is not specified, x86 is used as parameter to vcvars.bat (allowed on 64 bit systems, see https://msdn.microsoft.com/en-us/library/f2ccy3wt.aspx#Anchor_1).

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Thanks for the report @claasd 

As a regression, we will try to have a look as soon as possible, scheduled for next minor 1.9.1

**Comment by puetzk** [comment]
Msbuild honors an environment variable PreferredToolArchitecture=x64 as a way to set this: the exact logic in C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.Cpp.Common.props is

    
    <PreferredToolArchitecture Condition="'$(PreferredToolArchitecture)' == '' and ('$(UseNativeEnvironment)' == 'true' or '$(_IsNativeEnvironment)' == 'true')">x64</PreferredToolArchitecture>
    
    
    <PreferredToolArchitecture Condition="'$(PreferredToolArchitecture)' == '' and ('$(PROCESSOR_ARCHITECTURE)' == 'AMD64' and '$(Platform)' == 'x64')">x64</PreferredToolArchitecture>
    <PreferredToolArchitecture Condition="'$(PreferredToolArchitecture)' == ''">x86</PreferredToolArchitecture>
     
     
    <PreferredToolArchitecture Condition="'$(PreferredToolArchitecture)' == 'x64' and '$(PROCESSOR_ARCHITECTURE)' != 'AMD64' and '$(PROCESSOR_ARCHITEW6432)' != 'AMD64'">x86</PreferredToolArchitecture>

It might be logical for tools.vcvars_command to do the same if settings.arch_build is not set, rather than falling back to detected_architecture().

**Comment by memsharded** [comment]
If I understood it correctly I think the following logic should make it:
```python
    arch_setting = arch_setting or 'x86_64'
    arch_build = settings.get_safe("arch_build") or detected_architecture()
    if arch_build == 'x86_64':
        if settings.get_safe("arch_build") and int(compiler_version) >= 12:
            x86_cross = "amd64_x86"
        else:
            x86_cross = "x86"
        vcvars_arch = {'x86': x86_cross,
                       'x86_64': 'amd64',
                       'armv7': 'amd64_arm',
                       'armv8': 'amd64_arm64'}.get(arch_setting)
    elif arch_build == 'x86':
        vcvars_arch = {'x86': 'x86',
                       'x86_64': 'x86_amd64',
                       'armv7': 'x86_arm',
                       'armv8': 'x86_arm64'}.get(arch_setting)
```

**Comment by puetzk** [comment]
I agree that this restores the pre-1.7 behavior, though that behavior was pretty surprising. If you didn't list arch_build in settings, conan prefered amd64-hosted compilers except for arch=x86, where 1.7 didn't let you get an amd64-hosted compiler at all. This does let you reach it, but only if each package opts in by listing arch_build in settings. So we're back to there being no way for a profile to specify the amd64_x86 compiler if the package is ignorant of arch_build (e.g. almost all packages).

msbuild's defaults go the other way (default to x86-hosted compilers unless you've set `$(PreferredToolArchitecture)`, except that amd64 gets a native compiler by default if it's possible (i.e. if the $(PROCESSOR_ARCHITECTURE)' == 'AMD64'). I think it would make more sense to follow msbuild at least as far as honoring %PreferredToolArchitecture% (which you could set in a profile or using `conan create -e`).

And I'm not sure any of this will actually help @claasd, since the silverlight problem isn't about the VC toolchain at all, [it's about about msbuild itself](https://blogs.msdn.microsoft.com/visualstudio/2010/05/07/building-on-cross-targeting-scenarios-and-64-bit-msbuild/) (they try to load an x86-only dll from silverlight into the build system). I guess vcvars could be affecting the path in which msbuild.exe gets found, but for me `vcvarsall.bat amd64_x86`, `vcvarsall.bat amd64`, and `vcvarsall.bat x86` all result in finding C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\MSBuild.exe

**Comment by claasd** [comment]
Hi, 
basically, we need a way to ensure that vcvars.bat is called with x86 as parameter, them MsBuild loads the correct DLLs for Silverlight. I am open about how to achieve this.

**Comment by claasd** [comment]
@puetzk : When I read the props file correctly, the following defaults are used when PreferredToolArchitecture is not specified:
* set to x64 if UseNativeEnvironment or _IsNativeEnvironment is set to true
* else set to x64 if PROCESSOR_ARCHITECTURE equals "AMD64" and Platform equals "x64" (building for x64)
* else set to x86
* if os is x86, overwrite any previous defined x64 with x86.

So the default behavior would be to use x86 for this scenario, and set a variable PreferredToolArchitecture to "x64" if one wanted amd64_x86.

**Comment by lasote** [comment]
Hi, I think @puetzk is right about using `amd64_x86` even if no `setting.arch_build` is specified. If the detected architecture is `64` it should use `amd64_x86` (when supported), by default.

But of course, we have to enable to override the default. @claasd @memsharded What about to enabling a parameter `vcvars_arch()` in:  `vcvars_command()`, `vcvars_dict` and `MSBuild().build()` to override the calculation of the arch to be applied to `vcvars`? For me it is the common Conan behavior, the build helpers calculate things based on the settings, but any "calculation" should be allowed to be changed.

**Comment by claasd** [comment]
@lasote: No, it should NOT. When I specify x86 or Win32 in Visual Studio, by default it will build as if on an x86 system, even if I am on an x64 System. x86 should be the default, amd64_x86 should be the override.

**Comment by lasote** [comment]
The default of the Visual Studio doesn't necessarily need to follow the default of the `MSBuild()` build helper in conan.

**Comment by memsharded** [comment]
I think @puetzk confirm @claasd statements:

> default to x86-hosted compilers unless you've set $(PreferredToolArchitecture)

So, I'd say that the default is ``x86``, and then enable some way to opt-in into ``amd64_x86``

**Comment by memsharded** [comment]
I have updated https://github.com/conan-io/conan/pull/3918, to implement the following behavior:

- If defined or detected build architecture (``arch_build`` setting) is ``x86``, then the result will always be ``x86`` or ``x86_`` toolchains
- If defined or detected build architecture (``arch_build`` setting) is ``x86_64``:
   - If target arch (``arch`` setting) is `x86`:
      - If VS-version < 12 => ``x86`` (other not supported yet)
      - If user explicitly defined ``arch_build=64`` => ``amd64_x86``
      - If user explicitly defined env-var ``PreferredToolArchitecture=x64`` => ``amd64_x86``
      - Else =>  ``x86``

Please confirm @claasd and @puetzk that this behavior would be good for you.

```python
    arch_build = settings.get_safe("arch_build") or detected_architecture()
    if arch_build == 'x86_64':
        # Only uses x64 tooling if arch_build explicitly defines it, otherwise
        # Keep the VS default, which is x86 toolset
        # This will probably be changed in conan 2.0
        if ((settings.get_safe("arch_build") or
                os.getenv("PreferredToolArchitecture") == "x64")
                and int(compiler_version) >= 12):
            x86_cross = "amd64_x86"
        else:
            x86_cross = "x86"
        vcvars_arch = {'x86': x86_cross,
                       'x86_64': 'amd64',
                       'armv7': 'amd64_arm',
                       'armv8': 'amd64_arm64'}.get(arch_setting)
    elif arch_build == 'x86':
        vcvars_arch = {'x86': 'x86',
                       'x86_64': 'x86_amd64',
                       'armv7': 'x86_arm',
                       'armv8': 'x86_arm64'}.get(arch_setting)
```

Also, take into account that this will probably change in conan 2.0, specially if the cross-build settings model is updated, which is planned, and default to ``amd64_x86`` instead of ``x86``

**Comment by lasote** [comment]
> Also, take into account that this will probably change in conan 2.0, especially if the cross-build settings model is updated, which is planned, and default to amd64_x86 instead of x86

To clarify the motivation and explain the discussion we had in the Conan team: The rationale behind that pottential change in Conan 2.0 is that the Conan build helpers aim to be deterministic and not trust/follow any build system defaults, only the inputs. The `arch_build` (being the detected architecture, or the explicitly declared, or a new setting that Conan 2.0 could implement to improve the cross-build model) will be always used to determine the applied architecture. If finally in Conan 2.0 we use the detected architecture as an input to the build helper, it will determine the output unless it is overridden, without taking into account the build system defaults. Related issue #3919

**Comment by claasd** [comment]
Looks good to me.

**Comment by puetzk** [comment]
I think that's sane. I'd prefer that a profile's arch_build win out, but given how conan currently processing settings I don't see any way to implement it (the tools don't have access to the profile's full settings, only whatever the package listed).

And in the absence of a value from conan settings (at least in 1.x), I think following msbuild's default is in line with the principle of least astonishment. And a profile (or user) can set $(PreferredToolArchitecture) to their liking, which is at least *a* way to get there in 1.x, if a little messy.

**Comment by lasote** [comment]
Released in 1.9.1
